flair/rpc/rpc.go

```diff
@@ -545,7 +570,33 @@ func (s *server) Delete(ctx context.Context, req *ppb.DeleteRequest) (*ppb.Delet
 		return nil, status.Errorf(codes.InvalidArgument, "Invalid prefix provided: "+req.Prefix)
 	}
 	return &ppb.DeleteResponse{}, s.replicator.All(req.Prefix, func(srv *trie.Server) error {
+		ctx, cancel := context.WithTimeout(ctx, 10*s.timeout) // Multiply up timeout since this operation can be expensive.
+		defer cancel()
 		_, err := srv.GC.Delete(ctx, req)
 		return err
 	})
 }
+
+// withRollingTimeout implements a timeout between each step of a streamed RPC.
+// Essentially we want to fail if it takes > 5s (for example) to get a chunk, but we don't
+// want to put an upper limit on the total time the RPC can run for (it might be long
+// if it's a big file and things are moving kinda slowly).
+// It returns a new context which will be cancelled when the rolling timeout expired and
+// a callback to update its timeout.
+func withRollingTimeout(ctx context.Context, timeout time.Duration) (context.Context, func()) {
```

Comment:
I guess we have a general really long timeout set for gRPC in general (did you say 2 min). I suppose theoretically we could get a streaming endpoint sending us heartbeats while making no progress.

Reply:
I don't think these ones actually do that though; they should always be sending chunks of actual content.
I'm not too worried about one that is doing that, it's more if the server just vanishes halfway through the stream and we're getting literally nothing more - or to be able to time it out upfront if initially there's no response without limiting the whole stream in the same way.